doc/user-manual/language/core-language.lagda.rst

```diff
@@ -8,21 +8,182 @@
+
+Parser
+------
+
+.. _Happy: http://www.haskell.org/happy
+
+The parser is the component that takes the output of the lexer and builds a
+data structure that we will call Concrete Syntax, while checking for correct
+syntax.
+
+The parser is generated by Happy_.
+
+Example: when a name is a sequence of parts, the lexer just sees it as a
+string, the parser does the translation in this step.
+
+
+Concrete Syntax
+---------------
+
+The concrete syntax is a raw representation of the program text without any
+desugaring at all.  This is what the parser produces. The idea is that if we
+figure out how to keep the concrete syntax around, it can be printed exactly
+as the user wrote it.
+
+Nice Concrete Syntax
+--------------------
+
+The translation from "Agda.Syntax.Concrete" to "Agda.Syntax.Abstract" involves
+scope analysis, figuring out infix operator precedences and tidying up
+definitions.
+
+Abstract Syntax
+---------------
+
+The abstract syntax ``Agda.Syntax.Abstract`` is the result after desugaring
+and scope analysis of the concrete syntax. The type checker works on abstract
+syntax, producing internal syntax.
+
+Internal Syntax
+---------------
+
+This is the final stage of syntax before being handed off to one of the
+backends. Terms are well-scoped and well-typed.
+
+The internal syntax ``Agda.Syntax.Internal`` uses the following haskell
+datatype to represent a ``Term``.
 
 .. code-block:: haskell
 
-  data Term = Var Int Elims
-            | Def QName Elims               -- ^ @f es@, possibly a delta/iota-redex
-            | Con ConHead Args              -- ^ @c vs@
-            | Lam ArgInfo (Abs Term)        -- ^ Terms are beta normal. Relevance is ignored
-            | Lit Literal
-            | Pi (Dom Type) (Abs Type)      -- ^ dependent or non-dependent function space
-            | Sort Sort
-            | Level Level
-            | MetaV MetaId Elims
-            | DontCare Term
-              -- ^ Irrelevant stuff in relevant position, but created
-              --   in an irrelevant context.
+  data Term = Var {-# UNPACK #-} !Int Elims -- ^ @x es@ neutral
+          | Lam ArgInfo (Abs Term)        -- ^ Terms are beta normal. Relevance is ignored
+          | Lit Literal
+          | Def QName Elims               -- ^ @f es@, possibly a delta/iota-redex
+          | Con ConHead ConInfo Elims
+          -- ^ @c es@ or @record { fs = es }@
+          --   @es@ allows only Apply and IApply eliminations,
+          --   and IApply only for data constructors.
+          | Pi (Dom Type) (Abs Type)      -- ^ dependent or non-dependent function space
+          | Sort Sort
+          | Level Level
+          | MetaV {-# UNPACK #-} !MetaId Elims
+          | DontCare Term
+            -- ^ Irrelevant stuff in relevant position, but created
+            --   in an irrelevant context.  Basically, an internal
+            --   version of the irrelevance axiom @.irrAx : .A -> A@.
+          | Dummy String Elims
```

Comment:
`DontCare` and `Dummy` do not need to be shown in the user manual.

I suppose the internal syntax should be shown as a grammar rather than as Haskell code.

Reply:
OK, this was inherited from the old stub. I will update the grammar above.

Reply:
I rewrote the introductory sentence to the haskell term, I think it would be good if we show that it has the same elements as the grammar, what do you think?